pdg/src/info.rs

```diff
@@ -0,0 +1,515 @@
+use crate::graph::{Graph, GraphId, Graphs, Node, NodeId, NodeKind};
+use itertools::Itertools;
+use rustc_index::vec::IndexVec;
+use rustc_middle::mir::Field;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::cmp;
+use std::fmt::{self, Debug, Display, Formatter};
+
+/// The information checked in this struct is whether nodes flow to loads, stores, and offsets (pos
+/// and neg), as well as whether they are unique.
+/// Uniqueness of node X is determined based on whether there is a node Y which is X's ancestor
+/// through copies, fields, and offsets, also is a node Z's ancestor through copies, fields, offsets,
+/// where the fields are the same between X and Z, and where Z is chronologically between X and X's last descendent.
+/// If such a node exists, X is not unique.
+#[derive(Debug, Eq, PartialEq, Hash, Clone)]
+pub struct NodeInfo {
+    flows_to_mutation: Option<NodeId>,
+    flows_to_load: Option<NodeId>,
+    flows_to_pos_offset: Option<NodeId>,
+    flows_to_neg_offset: Option<NodeId>,
+    non_unique: Option<NodeId>,
+}
+
+//Algorithm for determining non-conflicting:
+//First calculate all of the below from the bottom-up.
+//Then create a tree where children are based on fields only,
+//each node having a list of NodeIds associated with it.
+//Finally, check that each node's starts and ends don't overlap with each other.
+#[derive(Debug, Hash, Clone, Copy)]
+struct GraphTraverseInfo {
+    last_descendent: NodeId,
+    flows_to_load: Option<NodeId>,
+    flows_to_store: Option<NodeId>,
+    flows_to_pos_offset: Option<NodeId>,
+    flows_to_neg_offset: Option<NodeId>,
+}
+
+fn init_traverse_info (n_id: NodeId, n: &Node) -> GraphTraverseInfo {
+    GraphTraverseInfo {
+        last_descendent: n_id,
+        flows_to_store: node_does_mutation(n).then(|| n_id),
+        flows_to_load: node_does_load(n).then(|| n_id),
+        flows_to_pos_offset: node_does_pos_offset(n).then(|| n_id),
+        flows_to_neg_offset: node_does_neg_offset(n).then(|| n_id)
+    }
+}
+
+fn create_flow_info (g: &Graph) -> HashMap<NodeId,GraphTraverseInfo> {
+    let mut f = HashMap::from_iter(
+        g.nodes.iter_enumerated()
+        .map(|(idx,node)| (idx,init_traverse_info(idx,node)))
+    );
+    for (n_id,node) in g.nodes.iter_enumerated().rev(){ 
+        let cur : GraphTraverseInfo = *(f.get(&n_id).unwrap());
+        if let Some(p_id) = node.source {
+            let parent = f.get_mut(&p_id).unwrap();
+            parent.last_descendent = cmp::max(cur.last_descendent,parent.last_descendent);
```

Comment:
Correct me if I'm missing something... since you're going in reverse order which is node id descending, I think it's enough to set this once iff it's not set, rather than doing a comparison

Reply:
it's also not a big deal if the above is true, what you have is simpler and gets the point across

Reply:
on the other hand, if you didn't go in reverse order, you could just set `parent.last_descendent = cur.last_descendent` without any checking, but that might require changes in the flow logic below. why do you traverse in reverse?

Reply:
> Correct me if I'm missing something... since you're going in reverse order which is node id descending, I think it's enough to set this once iff it's not set, rather than doing a comparison

If we have A, B, C, D, E, with A->B, A->C, B->E, C->D, then C will be encountered before B, and its last descendent is D, but the correct last descendent of A is E.

Reply:
ah, thanks for the counter-example